src/rgw/rgw_ioc_dispatch.cc

```diff
@@ -0,0 +1,270 @@
+      on_finish->complete(0);
+    }
+  });
+
+  Context* connect_ctx = new LambdaContext([this, register_ctx](int ret) {
+    if(ret < 0) {
+      ldout(m_cct, 1)  << "rgw_hook: IOC hook fail to connect to IOC daemon" << dendl;
+      register_ctx->complete(0);
+      return;
+    }
+
+    ldout(m_cct, 20) << "rgw_hook: IOC hook connected to IOC daemon." << dendl;
+
+    m_cache_client->register_client(register_ctx);
+  });
+
+
+	if(m_cache_client != nullptr && is_reconnect) {
+    delete m_cache_client;
+    auto controller_path = m_cct->_conf.template get_val<std::string>(
+            "immutable_object_cache_sock");
+    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
+	}
+
+	ceph_assert(m_cache_client != nullptr);
+
+  m_cache_client->run();
+  m_cache_client->connect(connect_ctx);
+}
+
+int IOChook::handle_register_client(bool reg) {
+  ldout(m_cct, 20) << "rgw_hook: registering client" << dendl;
+
+  if(!reg) {
+    ldout(m_cct, 1) << "rgw_hook: IOC hook register fails." << dendl;
+  }
+  return 0;
+}
+
+void IOChook::read(std::string oid, std::string pool_ns, int64_t pool_id,
+                   off_t read_ofs, off_t read_len, optional_yield y,
+                   rgw::Aio::OpFunc&& radosread, rgw::Aio* aio, rgw::AioResult& r
+                  ) {
+
+  /* if IOC daemon still don't startup, or IOC daemon crash,
+     or session occur any error, try to re-connect daemon. */
+  std::unique_lock locker{m_lock};
+  if(!m_cache_client->is_session_work()) {
+    // go to read from rados first
+    ldout(m_cct, 5) << "rgw_hook: session didn't work, go to rados" << oid << dendl;
+    std::move(radosread)(aio,r);
+    // then try to conect to IOC daemon.
+    ldout(m_cct, 5) << "rgw_hook: IOC hook try to re-connect to IOC daemon." << dendl;
+    create_cache_session(nullptr, true);
+    return;
+  }
+
+  CacheGenContextURef gen_ctx = make_gen_lambda_context<ObjectCacheRequest*,
+                                            std::function<void(ObjectCacheRequest*)>>
+            ([this, oid, read_ofs, read_len, y, radosread=std::move(radosread), aio, &r](ObjectCacheRequest* ack) mutable {
+    handle_read_cache(ack, oid, read_ofs, read_len, y, std::move(radosread), aio, r);
+  });
+
+	ceph_assert(m_cache_client != nullptr);
+  m_cache_client->lookup_object(pool_ns,
+                                pool_id,
+                                CEPH_NOSNAP,
+                                oid,
+                                std::move(gen_ctx));
+
+  return;
+}
+
+void IOChook::handle_read_cache(ObjectCacheRequest* ack, std::string oid,
+                                off_t read_ofs, off_t read_len, optional_yield y,
+                                rgw::Aio::OpFunc&& radosread, rgw::Aio* aio, rgw::AioResult& r
+                                ) {
+  if(ack->type != RBDSC_READ_REPLY) {
+    // go back to read rados
+    ldout(m_cct, 5) << "rgw_hook: IOC cache don't cache object, go to rados " << oid << dendl;
```

Comment:
maybe "object is not in cache" or "object not found in cache"

Reply:
resolved